### PR TITLE
Refactor SNative functionality

### PR DESCRIPTION
**Pull request by silasdavis**
As a precursor to #1241 

This work clears the ground for 'isotropic' calling between our
different execution engines: EVM, Natives, and WASM. In the process it
clears up a lot of cruft in an effort to make a better base for
implementing our current Tx/ExecutionContexts as natives, i.e.
'everything is a contract'.

In particular:

- Standardise on 'native' to describe former 'snatives' and introduce
package containing 'native' functionality including single
implementations for certain core shared functionality (like initialising code,
checking permissions, firing events). The idea is that some functionality
here can be wrapped up in native.Contract objects and be exposed like
the Permissions contract is now.

- Remove the evm.State interface and replace it with static helpers and
a simpler and generic (across VMs) CallFrame type that takes
responsibility for establishing disposal call frames and tracking stack
depth.

- Introduce the Callable interface and associated State and CallParams
types in a new 'engine' package. Make native functions and EVM contracts
implement Callable.

- Provide evm.Contract that is callable and use dispatcher so that EVM
contracts can call native contracts with the same semantics as other EVM
contracts. Via engine.Connect().

- Introduces Natives object that collects a selection of Callables and
passes external callables to them.

- Native callables are made to appear as if resident in state (TODO:
include them in backend state for real).

Signed-off-by: Silas Davis <silas@monax.io>

### CODE REVIEW

File: cmd/burrow/commands/snatives.go [hunk 1]
```diff
@@ -17,9 +17,9 @@ package commands
 import (
```
Comment (gregdhill) [inline cmd/burrow/commands/snatives.go]: Should we change the name of this file and command?

File: protobuf/acm.proto [hunk 2]
```diff
@@ -20,15 +20,25 @@ message Account {
...
     uint64 Sequence = 3;
+    // The account's current native token balance
     uint64 Balance = 4;
+    // We expect exactly one of EVMCode, WASMCode, and NativeName to be non-empty
+    // EVM bytecode
     bytes EVMCode = 5 [(gogoproto.customtype) = "Bytecode", (gogoproto.nullable) = false];
     permission.AccountPermissions Permissions = 6 [(gogoproto.nullable) = false];
+    // WASM bytecode
     bytes WASMCode = 7 [(gogoproto.customtype) = "Bytecode", (gogoproto.jsontag) = ",omitempty", (gogoproto.nullable) = false];
+    // Fully qualified name of native contract for which this Account is a sentinel
```
Comment (gregdhill) [inline protobuf/acm.proto]: I'm not sure I entirely understand this comment...

File: execution/errors/native.go [hunk 3]
```diff
@@ -11,10 +11,16 @@ type LacksSNativePermission struct {
 	SNative string
 }
 
-func (e LacksSNativePermission) Error() string {
+var _ CodedError = &LacksSNativePermission{}
+
+func (e *LacksSNativePermission) ErrorMessage() string {
```
Comment (gregdhill) [inline execution/errors/native.go]:
```suggestion
func (e *LacksNativePermission) ErrorMessage() string {
```

Comment (gregdhill) [reply execution/errors/native.go]: I think we should be consistent

File: execution/native/state.go [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+import (
+	"github.com/hyperledger/burrow/acm"
+	"github.com/hyperledger/burrow/acm/acmstate"
+	"github.com/hyperledger/burrow/binary"
+	"github.com/hyperledger/burrow/crypto"
+	"github.com/hyperledger/burrow/execution/errors"
+)
+
+// This wrapper provides a state that behaves 'as if' the natives were stored directly in state.
+// TODO: we may want to actually store native sentinels (and thier metadata) in on-disk state down the line
```
Comment (gregdhill) [inline execution/native/state.go]: I'm not exactly sure what you mean by 'sentinels'. Also typo

File: execution/native/native_test.go [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+	"fmt"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+)
+
+func TestDefaultNatives(t *testing.T) {
+	ns, err := DefaultNatives()
+	require.NoError(t, err)
+	fmt.Println(ns)
```
Comment (gregdhill) [inline execution/native/native_test.go]: Can we test this?

File: execution/execution_test.go [hunk 6]
```diff
@@ -1228,11 +1232,14 @@ func TestContractSend(t *testing.T) {
 	}
 
 	exe := makeExecutor(st)
+	fmt.Println("acc0", acc0.Address)
```
Comment (gregdhill) [inline execution/execution_test.go]: Let's test these for not nil

File: execution/native/account.go [hunk 7]
```diff
@@ -0,0 +1,218 @@
...
+	}
+	if acc.EVMCode != nil || acc.WASMCode != nil {
+		return errors.ErrorCodef(errors.ErrorCodeIllegalWrite,
+			"tried to initialise code for a contract that already has code: %v", address)
+	}
+
+	acc.EVMCode = code
+
+	// keccak256 hash of a contract's code
+	hash := sha3.NewKeccak256()
```
Comment (gregdhill) [inline execution/native/account.go]: Not necessarily something to be addressed in this PR, but I'm unsure why we maintain our own sha3 library in `crypto` when we can use the standard library? See: https://github.com/hyperledger/burrow/blob/master/crypto/hash.go#L11

File: execution/native/account.go [hunk 8]
```diff
@@ -0,0 +1,218 @@
...
+
+	acc.WASMCode = code
+	// keccak256 hash of a contract's code
+	hash := sha3.NewKeccak256()
+	hash.Write(code)
+	acc.CodeHash = hash.Sum(nil)
+	return st.UpdateAccount(acc)
+}
+
+func Transfer(st acmstate.ReaderWriter, from, to crypto.Address, amount uint64) error {
```
Comment (gregdhill) [inline execution/native/account.go]: Nice!

File: execution/native/call.go [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+	"github.com/hyperledger/burrow/execution/exec"
+)
+
+// Call provides a standard wrapper for implementing Callable.Call with appropriate error handling and event firing.
+func Call(state engine.State, params engine.CallParams, execute func(engine.State, engine.CallParams) ([]byte, error)) ([]byte, error) {
+	maybe := new(errors.Maybe)
+	if params.CallType == exec.CallTypeCall || params.CallType == exec.CallTypeCode {
+		// NOTE: DelegateCall does not transfer the value to the callee.
+
+		// DelegateCall is executed by the DELEGATECALL opcode, introduced as off Ethereum Homestead.
```
Comment (gregdhill) [inline execution/native/call.go]: Typo

File: protobuf/acm.proto (hunk 2 above)
Comment (silasdavis) [reply protobuf/acm.proto]: Fair. I'll expand in code, but what I meant:

- Fully qualified as in `Permissions.setBase` - I'm working on the assumption that every function has a globally unique name defined by the contract/function or just function if it is a 'loose' function
- Sentinel in the sense that this is an account object that represents a 'special' value, a native function, and doesn't contain enough information to fully process a call to the account (unlike EVM or WASM accounts. It's a marker account (at least for now). In the sense of : https://en.wikipedia.org/wiki/Sentinel_value

File: execution/native/native_test.go (hunk 5 above)
Comment (silasdavis) [reply execution/native/native_test.go]: this is probably debug, I think all I really want to test is the invocation not `String()`.

File: execution/native/account.go (hunk 7 above)
Comment (silasdavis) [reply execution/native/account.go]: I had no idea it existed in crypto - looks like it was committed 5th May 2018 so I guess that's why! I will change.

File: execution/execution_test.go (hunk 6 above)
Comment (silasdavis) [reply execution/execution_test.go]: I will fix this, though these are just stray debug lines.

Alas there are quite a few unchecked errors in this test. Most of which are guaranteed not to be errors in the context of these tests but still that should be made explicit. execution_test needs a complete rework. I will open a ticket for this.

**Pull request merged**